### PR TITLE
Improve performance for Groups with many EPerson members. Fix pagination on endpoints

### LINKED ISSUE #9052: Performance Issues with Groups having many EPerson members
**Describe the bug**
I'm currently utilizing DSpace 7.5, where I've encountered a challenge. Within my system, there are approximately 1000 groups and 2000 members as epersons. One particular group contains 1300 members (epersons), while others have 50-100 members(epersons). When I attempt to search for the group with 1300 members (epersons) and one subgroup,  it takes a considerable amount of time to retrieve the results and when I click on edit it also takes 4-5 minutes to retrieve the current members (persons). I've also conducted tests using hal browser and found that the response time is significantly delayed.

The URL I use for this search is as follows:
`/server/api/eperson/groups/6d2c20f0-f1b1-4ce2-b613-8c6816e9b90a/epersons
`
Even when I limit the search to just one result with the **size=1** parameter:
`/server/api/eperson/groups/6d2c20f0-f1b1-4ce2-b613-8c6816e9b90a/epersons?page=0&size=1`, it still takes 3-4 minutes to receive a response. However, when I search for other groups with fewer members (50-100), the system performs smoothly.

I should mention that this issue is occurring in my local environment, using localhost:8080 and localhost:4000.

**To Reproduce**
1. Go to Access Control and click on Groups.
2. Then the Search group name in my case I have searched Students then it takes more than 4-5 minutes to retrieve the group(student).
3. After clicking on edit. again it takes 4-5 minutes to show Current Members.

**Expected behavior**
Since my other groups also require less time, it should consequently require less time.

### PR DESCRIPTION
**Pull request by tdonohue**
## References
* Fixes #9052 by providing major speed improvements on backend & proper pagination of group members.
* _NOTE:_ This is not a complete fix, but will improve performance significantly.  Additional endpoints will be necessary to completely fix performance issues in Angular UI.  See https://github.com/DSpace/dspace-angular/issues/2512

## Description
This PR improves the performance of Groups which have many EPerson (or Group) members by adding proper pagination to these endpoints. (Previously all objects were always loaded into memory _before_ pagination was applied.)
* [`GET /api/eperson/groups/<:uuid>/epersons`](https://github.com/DSpace/RestContract/blob/main/epersongroups.md#get-epeople-for-a-single-eperson-group)
* [`GET /api/eperson/groups/<:uuid>/subgroups`](https://github.com/DSpace/RestContract/blob/main/epersongroups.md#get-sub-groups-in-a-single-parent-eperson-group)

This was implemented via new paginated methods in EPersonService/DAO and GroupService/DAO:
* EPersonService/DAO.findByGroups() -> Locates all EPersons who are a member of one or more groups in a paginated manner.
* EPersonService/DAO.countByGroups() -> Provides total count of those EPerson members (necessary for pagination)
* GroupService/DAO.findByParent() -> Locates all (Sub)Groups who are a member group of the given Parent group in a paginated manner.
* GroupService/DAO.countByParent() -> Provides total count of those (Sub)Groups members (necessary for pagination)
* Unit & Integration Tests were added for these new methods

Warnings were added to JavaDoc for the following methods. All of which will have very BAD performance for large groups:
* `Group.getMembers()`
* `Group.getMemberGroups()`
* `EPersonService.findByGroups()` (unpaginated version)
* `GroupService.allMembers()`
* NOTE: I did NOT remove all usages of these methods.  Attempting to remove all usages of these methods would require more discussion/thought. I just replaced the simple ones (i.e. some calls to `allMembers()`) as described below.

Finally, I discovered several usages of `GroupService.allMembers()` where the code only required the `size()` of that list.  These usages were replaced with new `countAllMembers()` or `countByParent()` methods. All these usages have existing detailed tests were still pass with the newly refactored code:
* `EPersonService.delete(Context, EPerson, cascade)`
* `GroupService.removeMember(Context, Group, EPerson)`
* `GroupService.removeMember(Context, Group, Group)`

## Instructions for Reviewers
* See #9052. 
* In UI, verify "Access Control" -> Groups is working properly for search/editing.
* In REST API, verify pagination still works properly for BOTH the `/epersons` and `/subgroups` endpoints (size = how many to return, page = which page of results to return).  With this PR in place, small pages (small values of `size`) will return a response quickly.  Without this PR (e.g. in 7.6), the `size` of page doesn't matter as every object is loaded into memory regardless of `size` before a response is returned.
    * e.g. `GET /api/eperson/groups/<:uuid>/epersons?page=0&size=1`  (Only return one result at a time per page)
    * e.g. `GET /api/eperson/groups/<:uuid>/subgroups?page=0&size=5` (Return first page of 5 results per page)
* Review code and new tests
* Verify all automated tests pass.

### CODE REVIEW

**Comment by tdonohue** [comment]
@alanorth : Thanks for the tests!  I added more details under "Instructions for Reviewers" on how to test pagination.  If you are interested, It can be done either via commandline (wget or curl) or from the HAL Browser.  (Pagination also has some automated tests in my updates to `GroupRestRepositoryIT` in this PR, so I feel confident it works.  But that's how you'd manually test it)

**Comment by tdonohue** [comment]
@alanorth : Apologies, I've misworded the description slightly.  Yes, paging *already* works in 7.6... but this PR changes the paging to provide better performance.  In 7.6, paging will first load every object into memory (in your test, 512 objects will be loaded into memory, as `totalElements=512`) before returning just the few necessary for the page.   With this PR in place, *only the objects on the page (i.e. `size` value) will be loaded into memory.

So, if you test with 7.6, you'll see slower behavior to the paging. As every page will load 512 objects before returning the response (even when `size=1`).  But, if you test with this PR, you'll see much quicker responses, as you'll only load 1 object (when `size=1`).

It can be difficult to see the speed difference on smaller repositories, but it should still be noticeable.

**Comment by alanorth** [comment]
@tdonohue ok thanks for taking the time to explain. I was contemplating not commenting, assuming I was doing something stupid. So I am +1 on this by testing because it doesn't change functionality at all and, while I didn't notice the performance increase on localhost in my case, it didn't have any negative affect either.

I will let someone else review the code since I don't know it well.

**Review by vins01-4science — CHANGES_REQUESTED** [review]
Hello @tdonohue , thank you very much for this improvement :rocket: !

However, I am a little bit concerned about the paginated queries that are not using any ordering. I think that in both cases we should offer an order, otherwise the pagination cannot be used in a real scenario. 

On the other hand I found out some other little improvements that can be made, and that I suggest that can be done.

File: dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java [hunk 1]
```diff
@@ -179,8 +179,13 @@ public void removeMember(Context context, Group group, EPerson ePerson) throws S
...
                             && claimedTask.getWorkflowItem().getCollection() == collectionRole.getCollection()) {
-                        List<EPerson> ePeople = allMembers(context, group);
-                        if (ePeople.size() == 1 && ePeople.contains(ePerson)) {
+                        // Count number of EPersons who are *direct* members of this group
+                        int totalDirectEPersons = ePersonService.countByGroups(context, Set.of(group));
+                        // Count number of Groups which have this groupParent as a direct parent
+                        int totalChildGroups = countByParent(context, group);
+                        // If this group has only one direct EPerson and *zero* child groups, then we cannot delete the
+                        // EPerson or we will leave this group empty.
+                        if (totalDirectEPersons == 1 && totalChildGroups == 0) {
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:188]: @vins01-4science : I believe you misunderstand what the code is doing here.  It is blocked from being called multiple times by the `if` on line 180.  As that `if` ensures that we only run these two "count" queries **if** we find that the EPerson has a ClaimedTask where that ClaimedTasks' workflow process relates back to a Collection-specific Group.  It should never run multiple times, but is placed in this location to avoid running the count queries *except* where that `if` on line 180 passes.

I also need to point out, I didn't write this method in this PR.  I simply modified the verification logic.  So, I don't think we should expect this PR to cleanup code which is *not* part of the PR.  I do agree with you that this entire `removeMember()` method could be broken down better and maybe written better to avoid nested for loops...but that is not logic that I added in this PR, so I shouldn't be expected to fix someone else's code.  I don't see an easy way to remove the nested `for` loops without a very large code refactor, and I don't think we should require that in this PR.

In summary, in this method, all I did was replace a very poor performing `allMembers()` method with two `countBy` methods that are much faster.  I didn't write the rest of the logic here.

File: dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java [hunk 2]
```diff
@@ -191,8 +196,13 @@ public void removeMember(Context context, Group group, EPerson ePerson) throws S
...
             if (!poolTasks.isEmpty()) {
-                List<EPerson> ePeople = allMembers(context, group);
-                if (ePeople.size() == 1 && ePeople.contains(ePerson)) {
+                // Count number of EPersons who are *direct* members of this group
+                int totalDirectEPersons = ePersonService.countByGroups(context, Set.of(group));
+                // Count number of Groups which have this groupParent as a direct parent
+                int totalChildGroups = countByParent(context, group);
+                // If this group has only one direct EPerson and *zero* child groups, then we cannot delete the
+                // EPerson or we will leave this group empty.
+                if (totalDirectEPersons == 1 && totalChildGroups == 0) {
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:205]: Same as the above comment, I didn't write this method.  All I did was remove a poor performing `allMembers()` call with two (much faster) `countBy` method calls.  I don't think I should be required to refactor all the code around this small change. 

That said, I fully agree this code is messy.  I just won't have time to pull it apart and rewrite it, and that is a much larger change than I was attempting in this PR.

File: dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java [hunk 3]
```diff
@@ -212,9 +222,13 @@ public void removeMember(Context context, Group groupParent, Group childGroup) t
...
-                List<EPerson> parentPeople = allMembers(context, groupParent);
-                List<EPerson> childPeople = allMembers(context, childGroup);
-                if (childPeople.containsAll(parentPeople)) {
+                // Count number of Groups which have this groupParent as a direct parent
+                int totalChildGroups = countByParent(context, groupParent);
+                // Count number of EPersons who are *direct* members of this group
+                int totalDirectEPersons = ePersonService.countByGroups(context, Set.of(groupParent));
+                // If this group has only one childGroup and *zero* direct EPersons, then we cannot delete the
+                // childGroup or we will leave this group empty.
+                if (totalChildGroups == 1 && totalDirectEPersons == 0) {
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:231]: Same as the above two comments... Just replacing a poor performing `allMembers()` call with two better performing `countBy` calls.  This avoids loading every single EPerson object into memory for these two groups.

File: dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java [hunk 4]
```diff
@@ -381,6 +395,21 @@ public List<EPerson> allMembers(Context c, Group g) throws SQLException {
         return new ArrayList<>(childGroupChildren);
     }
 
+    @Override
+    public int countAllMembers(Context context, Group group) throws SQLException {
+        // Get all groups which are a member of this group
+        List<Group2GroupCache> group2GroupCaches = group2GroupCacheDAO.findByParent(context, group);
+        Set<Group> groups = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:402]: I'm not sure I agree.  With HashSets, the size will automatically double when it reaches 3/4 capacity.  So, if I set the capacity to the full list size, my understanding is it will double in size _before_ I finish adding the list (making it much larger than I need).  I think that can negatively impact our performance here.

See discussion here https://stackoverflow.com/a/38006602

Overall, I've always found/learned that presetting a HashSet's size doesn't help much.  But, maybe you've seen something that I've overlooked?

File: dspace-api/src/main/java/org/dspace/eperson/dao/EPersonDAO.java [hunk 5]
```diff
@@ -38,7 +38,29 @@ public List<EPerson> search(Context context, String query, List<MetadataField> q
 
     public int searchResultCount(Context context, String query, List<MetadataField> queryFields) throws SQLException;
 
-    public List<EPerson> findByGroups(Context context, Set<Group> groups) throws SQLException;
+    /**
+     * Find all EPersons who are a member of one or more of the listed groups in a paginated fashion. Order is
+     * indeterminate.
```
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/eperson/dao/EPersonDAO.java:43]: I think you are correct here. I think the order *is* consistent, but it may order by UUID.  I'll dig into this more and make sure to fix the docs and/or code here.

File: dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vins01-4science) [reply dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java:402]: Yeah, I missed the loadfactor in the count, it should be: `new HashSet<>((group2GroupCaches.size()/0.75) + 1);`
Generally you are right about `HashMaps`, but since in this case we are goingo to only populate it using a fixed number of elements, it will be better to declare optimal size at the start.
I think that is high probably to have more than 12 elements (default size is 16, with loadfactor of 0.75 (3/4)), so at least it will be recreated one time, this will decrease performance every time it will be recreated (by doubling its size).
So if we use an optimal init size, since we are using it only to store and query elements, the map won't be resized during the insertions, and all the read operations will be optimal.
Some reference:
- https://stackoverflow.com/a/21822273

File: dspace-api/src/main/java/org/dspace/eperson/dao/EPersonDAO.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/eperson/dao/EPersonDAO.java]: I've verified this comment is incorrect.  The default ordering of findByGroups() is by EPerson UUID.  I didn't modify this behavior in this PR, but have corrected the comment in https://github.com/DSpace/DSpace/pull/9078/commits/09e7f105e063ffa7fa35abe7cd9160c59b0f91a8.  This PR just added pagination to findByGroups() without modifications to the underlying query itself.

File: dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tdonohue) [reply dspace-api/src/main/java/org/dspace/eperson/GroupServiceImpl.java]: Good point. I've corrected this (and another instance of HashSet in the same class) in this latest commit: https://github.com/DSpace/DSpace/pull/9078/commits/09e7f105e063ffa7fa35abe7cd9160c59b0f91a8

**Comment by tdonohue** [comment]
@vins01-4science : Thanks for your review.  I've addressed all feedback in the latest commit.  That said, I did NOT fix the issues you pointed out with nested `for` loops in `GroupServiceImpl` (and similar) as this PR didn't create that code.  While I agree that code could be cleaned up better, I won't have time to refactor it for the 7.6.1 release... so I'd prefer moving forward with this PR as-is and only reviewing the code this PR adds/removes.  Thanks!

**Review by vins01-4science — APPROVED** [review]
Looks good to me, thank you @tdonohue :rocket:

**Comment by tdonohue** [comment]
Merging as this is at +2.
